midend/parserUnroll.cpp

```diff
@@ -147,19 +147,19 @@ class ParserStateRewriter : public Transform {
             BUG_CHECK(l->is<SymbolicArray>(), "%1%: expected an array", l);
             auto array = l->to<SymbolicArray>();
             unsigned idx = 0;
-            for (size_t i = 0; i < array->size; i++) {
-                auto* v = array->get(expression, i);
-                if (v->hasUninitializedParts())
-                    break;
-                else
-                    idx = i;
+            unsigned offset = 0;
+            if (state->statesIndexes.count(expression->expr->toString())) {
+                idx = state->statesIndexes.at(expression->expr->toString());
+                if (idx + 1 < array->size && expression->member.name != IR::Type_Stack::last) {
+                    offset = 1;
+                }
             }
             if (expression->member.name == IR::Type_Stack::lastIndex) {
                 return new IR::Constant(IR::Type_Bits::get(32), idx);
             } else {
-                state->statesIndexes[expression->expr->toString()] = idx;
+                state->statesIndexes[expression->expr->toString()] = idx + offset;
```

Comment:
`expression->expr->toString()` is there no better way to map this?

Reply:
clone_id can't be applied here. It is possible to implement operator< for IR::Member, but IR::Member::member still contains cstring. So, for now, I think that this is the best solution.

Reply:
clone_id can't be applied here. It is possible to implement operator< for IR::Member, but IR::Member::member still contains cstring. So, for now, I think that this is the best solution.

Reply:
node->toString() should only be used for reporting user messages.
Such usage in the compiler is unsafe, since the different expressions may generate identical toString outputs.